### PR TITLE
Avoid colliding names when components have enclosed types.

### LINKED ISSUE #98: Generated component when there's a name collision between nested type and provided type
Compilation of a generated component currently fails if a provided type has the same simple name as a nested type of the component, or that of its supers.

Example which shows a forged example, clashing on Buffer. 

```java
@dagger.Component(modules = FooComponent.Module.class)
public interface FooComponent {

  class Buffer {
  }

  @dagger.Module
  public static final class Module {
    @dagger.Provides java.nio.Buffer internal() {
      return null;
    }

    @dagger.Provides Buffer buffer(java.nio.Buffer internal) {
      return new Buffer();
    }
  }

  Buffer buffer();
}
```

The reason the generated type fails compilation is that it doesn't fully qualify the type parameter of the provider.

Ex.

```java
  private Provider<Buffer> internalProvider;
```

which should be fully qualified like so:

```java
  private Provider<java.nio.Buffer> internalProvider;
```

### PR DESCRIPTION
**Pull request by adriancole**
fixes #98, but probably @JakeWharton will have some things to say on the implementation approach.

### CODE REVIEW

File: compiler/src/test/java/dagger/internal/codegen/writer/JavaWriterTest.java [hunk 1]
```diff
@@ -15,19 +15,64 @@
...
 @RunWith(JUnit4.class)
 public class JavaWriterTest {
-  @Test public void referencedAndDeclaredSimpleName() {
+
+  @Test public void referencedAndDeclaredSimpleName() throws IOException {
     JavaWriter javaWriter = JavaWriter.inPackage("test");
     ClassWriter topClass = javaWriter.addClass("Top");
     topClass.addNestedClass("Middle").addNestedClass("Bottom");
     topClass.addField(ClassName.create("some.other.pkg", "Bottom"), "field");
-    assertThat(topClass.toString()).doesNotContain("import some.other.pkg.Bottom;");
```
Comment (adriancole) [inline compiler/src/test/java/dagger/internal/codegen/writer/JavaWriterTest.java]: test bug. toString doesn't generate code.

**Comment by adriancole** [comment]
I misunderstood originating elements. I'll have to rework this to decouple from that. The work I am doing is dependent on bugfixes in https://github.com/google/dagger/pull/96 anyway, so hopefully we can get that merged sometime.

**Comment by cgruber** [comment]
Hrm.  This P/R is further out of date than I'd like - I'm really sorry Adrian.  We've been doing so much churn internally, that it's made taking pulls externally much harder.  The code should stabilize a lot going forward.

I'm consulting with @gk5885 to see if we want to update this P/R and take it, or if he's got a fix in mind.

**Comment by adriancole** [comment]
I can rebase if you like, or even close it out. It has indeed been a while..

**Comment by cgruber** [comment]
I'm waiting to hear from greg about related commits internally.  I'll close it if it's handled elsewise, or comment more here.

**Comment by cgruber** [comment]
Hey @adriancole.  I realize we've been pokey and you've been attentive, but if you're able to update this quickly, we can get it into 2.0.  If not, I suspect we have a 2.0.1 coming shortly after to deal with minor improvements, and 2.1 not too far off with some API lock down and more important improvements.

**Comment by cgruber** [comment]
Tentatively putting this in the 2.0.1 bucket.  It's a good change, and we should put it in as soon as we can.

**Comment by adriancole** [comment]
ok on it. probably will have this passing tests again tomorrow.

File: compiler/src/test/java/dagger/internal/codegen/ComponentProcessorTest.java [hunk 2]
```diff
@@ -1837,4 +1837,274 @@ public void genericTestToLetMeDebugInEclipse() {
...
+        "import java.nio.Buffer;",
+        "import javax.annotation.Generated;",
+        "import javax.inject.Provider;",
+        "import test.ComponentWithEnclosedType;",
+        "",
+        "@Generated(\"dagger.internal.codegen.ComponentProcessor\")",
+        "public final class ComponentWithEnclosedType$Buffer_Factory implements Factory<ComponentWithEnclosedType.Buffer> {",
+        "  private final Provider<Buffer> bProvider;",
+        "",
+        "  public ComponentWithEnclosedType$Buffer_Factory(Provider<Buffer> bProvider) {",
```
Comment (adriancole) [inline compiler/src/test/java/dagger/internal/codegen/ComponentProcessorTest.java]: curious.. are we intentionally still having these ctors public eventhough there's a factory?

**Comment by adriancole** [comment]
rebasing helped catch me up on how the compiler generates files now :)

**Comment by cgruber** [comment]
Ok - sorry - we're just catching back up to prepare for a 2.0.1.  Is this in a shape that you feel is ready?

**Comment by cgruber** [comment]
FYI, we're shooting for thursday.

**Comment by cgruber** [comment]
Hey @adriancole - did you still want to do this, or should we re-do this internally?

**Comment by adriancole** [comment]
sorry. was out of town. will rebase tomorrow morning!

**Comment by adriancole** [comment]
ok rebased!

**Comment by ronshapiro** [comment]
I think we should close this in favor of https://github.com/square/javapoet/issues/77. #98 should still be kept open since the issue still arises, but we've cut out JavaWriter and solving this just for all JavaPoet users is probably the best thing to do. @adriancole is that ok?

**Pull request closed without merging**